### ISSUE TITLE
risc0-build fails to rename a downloaded file across different filesystems

### ISSUE THREAD

**Issue by fishilico** [risc0/risc0#496]
Hello,

While testing the first tutorial on https://www.risczero.com/docs/examples/hello_multiply , `cargo build` failed with the following stack trace:

```text
root@70e42d1e0e4f:/host/test_project# RUST_BACKTRACE=1 cargo build
   Compiling methods v0.1.0 (/host/test_project/methods)
error: failed to run custom build command for `methods v0.1.0 (/host/test_project/methods)`

Caused by:
  process didn't exit successfully: `/host/test_project/target/debug/build/methods-01681353818bdf03/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=RISC0_SKIP_BUILD
  Standard library /host/test_project/target/debug/build/methods-1e7d25267347a754/out/rust-std_4f29d950e101936e934eafe043bfaa02f991c74575a6dfc8647c3c7fbbdede58 does not exist; downloading
  Downloading https://github.com/risc0/rust/archive/53bbc8fc2afb2e10e3a90d7bf188bfd6598374ab.zip, mapping rust-53bbc8fc2afb2e10e3a90d7bf188bfd6598374ab/library to /host/test_project/target/debug/build/methods-1e7d25267347a754/out/rust-std_4f29d950e101936e934eafe043bfaa02f991c74575a6dfc8647c3c7fbbdede58.downloadtmp/library
  Downloaded: /host/test_project/target/debug/build/methods-1e7d25267347a754/out/.tmp99kwTn/53bbc8fc2afb2e10e3a90d7bf188bfd6598374ab.zip

  --- stderr
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 18, kind: CrossesDevices, message: "Invalid cross-device link" }', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/risc0-build-0.14.0/src/lib.rs:301:70
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:575:5
     1: core::panicking::panic_fmt
               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/panicking.rs:65:14
     2: core::result::unwrap_failed
               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/result.rs:1791:5
     3: core::result::Result<T,E>::unwrap
               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/result.rs:1113:23
     4: risc0_build::download_zip_map
               at /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/risc0-build-0.14.0/src/lib.rs:301:13
     5: risc0_build::setup_guest_build_env
               at /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/risc0-build-0.14.0/src/lib.rs:244:9
     6: risc0_build::embed_methods_with_options
               at /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/risc0-build-0.14.0/src/lib.rs:459:27
     7: risc0_build::embed_methods
               at /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/risc0-build-0.14.0/src/lib.rs:511:5
     8: core::ops::function::FnOnce::call_once
               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/ops/function.rs:251:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The line which panics is in function `download_zip_map` of `risc0-build` https://github.com/risc0/risc0/blob/c657ca0747947db4b890751853c2d3358d820f31/risc0/build/src/lib.rs#L301

To debug things, I printed the parameters of `fs::rename`:

```rust
println!("Trying to rename {:?} to {:?}", temp_dir.path().join(zm.filename), zip_path);
```

This printed:

```text
  Trying to rename "/host/test_project/target/debug/build/methods-d88b85b8ae1739c5/out/.tmpWWuFkr/53bbc8fc2afb2e10e3a90d7bf188bfd6598374ab.zip"
to "/root/.cache/risc0/53bbc8fc2afb2e10e3a90d7bf188bfd6598374ab.zip"
```

I am using a container (on Ubuntu 22.04) to build this project, where `/host` is shared with my main filesystem and `/root/.cache` a directory in the (temporary) container filesystem. So `fs::rename` fails, and this is normal according to the Rust documentation (https://doc.rust-lang.org/std/fs/fn.rename.html states that it fails when "`from` and `to` are on separate filesystems.").

**TL;DR** Function `download_zip_map` of `risc0-build` wants to rename a temporary file from the output directory to the user cache and this fails when these locations use different filesystems.

I see at least two possible way to fix this issue: use a temporary file in the cache directory (so that `fs::rename` also does not cross any directory) ; or catch errors and fall back to a copy+delete implementation of the move. What do you think?

**Comment by flaub** [comment]
The reason a rename is used instead of copy/delete is that a rename is atomic. This is important because cargo may run multiple builds in parallel.

Perhaps what we could do is allow users to configure where the cache directory is, and add some documentation that states that this directory must be on the same filesystem as the output directory. Ideally we'd put the cache directory into the output directory, the downside being that if someone deletes the output dir, the cache is also blown away.

**Comment by fishilico** [comment]
This seems reasonable. Thanks for your quick reply!

Anyway, I did not understand why the temporary download directory (`temp_dir`) is currently being created in the output directory instead of the cache, in https://github.com/risc0/risc0/blob/c657ca0747947db4b890751853c2d3358d820f31/risc0/build/src/lib.rs#L272

If it is possible to use a subdirectory of the cache as the temporary download directory, the issue can also be solved by using `cache_dir`, for example with:
```rust
let temp_dir = tempdir_in(cache_dir).unwrap();
```

**Comment by flaub** [comment]
@mothran Any objections to moving the temp dir to be inside the cache dir?

**Comment by mothran** [comment]
Nope! Should be a good solution IMHO.

**Comment by mothran** [comment]
@fishilico if you would like to test this fix: https://github.com/risc0/risc0/pull/497 that would be helpful.

I think you can do so by just redirecting the `risc0-build` dependency in `methods/Cargo.toml` and `./methods/guest/Cargo.toml` your project via:
```
risc0-build = { git = "https://github.com/risc0/risc0.git", branch = "parker/cache-tmpdir" }
```

**Comment by mothran** [comment]
> @fishilico if you would like to test this fix: #497 that would be helpful.
> 
> I think you can do so by just redirecting the `risc0-build` dependency in `methods/Cargo.toml` and `./methods/guest/Cargo.toml` your project via:
> 
> ```
> risc0-build = { git = "https://github.com/risc0/risc0.git", branch = "parker/cache-tmpdir" }
> ```

Sorry change that to `branch = "main"` after we just merged the fix.

**Comment by fishilico** [comment]
> @fishilico if you would like to test this fix: #497 that would be helpful.
> 
> I think you can do so by just redirecting the `risc0-build` dependency in `methods/Cargo.toml` and `./methods/guest/Cargo.toml` your project via:
> 
> ```
> risc0-build = { git = "https://github.com/risc0/risc0.git", branch = "parker/cache-tmpdir" }
> ```

Hello, I have tested this (with `branch = "main"`) and I confirm it fixed the issue I had. Thanks!
